### PR TITLE
Add parser option for parsing SQL numeric literals as decimal

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/4072

# Rationale for this change

Main motivation is to get benchmark query q6 returning the correct results but this is also the correct behavior for SQL

# What changes are included in this PR?

- Add config to enable parsing numbers as decimal. We cannot enable by default until decimal support is farther along

# Are there any user-facing changes?

### CODE REVIEW

**Comment by andygrove** [comment]
@kmitchener @viirya @Dandandan PTAL

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -2671,6 +2694,49 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            Ok(lit(n))
+        } else if self.options.parse_float_as_decimal {
+            if let Some(i) = n.find('.') {
+                let p = n.len() - 1;
+                let s = n.len() - i - 1;
+                let str = n.replace('.', "");
+                let n = str.parse::<i128>().map_err(|_| {
+                    DataFusionError::from(ParserError(format!(
+                        "Cannot parse {} as i128 when building decimal",
+                        str
```
Comment (viirya) [inline datafusion/sql/src/planner.rs:2717]: Maybe good to print out original string too, i.e.,

```suggestion
                        "Cannot parse {} as i128 when building decimal. Input token: {}",
                        str, n
```

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -2907,28 +2973,26 @@ pub fn convert_data_type(sql_type: &SQLDataType) -> Result<DataType> {
...
     use std::any::Any;
 
+    #[test]
+    fn parse_decimals() {
+        let options = ParserOptions {
+            parse_float_as_decimal: true,
+        };
+        quick_test_with_options(
+            "SELECT 1, 1.0, 0.1, .1, 12.34",
+            "Projection: Int64(1), Decimal128(Some(10),2,1), Decimal128(Some(1),2,1), Decimal128(Some(1),1,1), Decimal128(Some(1234),4,2)\
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply datafusion/sql/src/planner.rs:2990]: agree

[the remark this answers is not in the document]
Comment (liukun4515) [reply datafusion/sql/src/planner.rs:2990]: we should try to use the min precision to represent the value in the decimal literal

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -2671,6 +2694,49 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                    s as u8,
+                )))
+            } else {
+                let number = n.parse::<i128>().map_err(|_| {
+                    DataFusionError::from(ParserError(format!(
+                        "Cannot parse {} as i128 when building decimal",
+                        n
+                    )))
+                })?;
+                Ok(Expr::Literal(ScalarValue::Decimal128(Some(number), 38, 0)))
```
Comment (viirya) [inline datafusion/sql/src/planner.rs:2732]: Can we get the minimum precision to handle this number?


```
scala> sql("select 10000000000000000000")
res10: org.apache.spark.sql.DataFrame = [10000000000000000000: decimal(20,0)]
```

Comment (andygrove) [reply datafusion/sql/src/planner.rs:2730]: @viirya Could you take another look when you have time?

Comment (viirya) [reply datafusion/sql/src/planner.rs:2730]: I'd love to do it. Let me take another look.

File: datafusion/sql/src/planner.rs [hunk 4]
```diff
@@ -2919,19 +2987,103 @@ fn extract_possible_join_keys(
...
+        | SQLDataType::CharLargeObject(_)
+        | SQLDataType::Clob(_) => Err(DataFusionError::NotImplemented(format!(
+            "Unsupported SQL type {:?}",
+            sql_type
+        ))),
+    }
+}
+
+/// Convert SQL data type to relational representation of data type
+pub fn convert_data_type(sql_type: &SQLDataType) -> Result<DataType> {
```
Comment (viirya) [inline datafusion/sql/src/planner.rs:3076]: Is this committed in accidentally?

Comment (viirya) [reply datafusion/sql/src/planner.rs:3076]: Not sure why this change adds this and `convert_simple_data_type`. Otherwise other change looks good to me.

Comment (andygrove) [reply datafusion/sql/src/planner.rs:3076]: Yeah, this is from a merge conflict. I will fix.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
🎉

**Pull request merged**